kythe/go/platform/tools/kzip/viewcmd/viewcmd.go

```diff
@@ -0,0 +1,169 @@
+package viewcmd // import "kythe.io/kythe/go/platform/tools/kzip/viewcmd"
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/subcommands"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+	"kythe.io/kythe/go/platform/kzip"
+	"kythe.io/kythe/go/util/cmdutil"
+)
+
+type cmd struct {
+	cmdutil.Info
+	extractDir string
+}
+
+// New returns an implementation of the "view" subcommand.
+func New() subcommands.Command {
+	const usage = `Usage: view [options] <file-path>...
+
+Print or extract compilation units stored in .kzip files.  By default, the
+compilation unit is printed to stdout in JSON format.
+
+With -extract the compilation record and the full contents of all the required
+input files are extracted into the named directory, preserving the path
+structure specified in the compilation record.`
+
+	return &cmd{
+		Info: cmdutil.NewInfo("view", "view compilation units stored in .kzip files", usage),
+	}
+}
+
+// SetFlags implements part of subcommands.Command.
+func (c *cmd) SetFlags(fs *flag.FlagSet) {
+	fs.StringVar(&c.extractDir, "extract", "", "Extract files to this directory")
+}
+
+// Execute implements part of subcommands.Command.
+func (c *cmd) Execute(ctx context.Context, fs *flag.FlagSet, args ...interface{}) subcommands.ExitStatus {
+	hasErrors := false
+	for _, path := range fs.Args() {
+		ext := filepath.Ext(path)
+		base := filepath.Base(strings.TrimSuffix(path, ext))
+
+		f, err := os.Open(path)
+		if err != nil {
+			log.Printf("Error opening .kzip file: %v", err)
+			hasErrors = true
+			continue
+		}
+		defer f.Close()
+
+		err = kzip.Scan(f, func(r *kzip.Reader, unit *kzip.Unit) error {
+			if err := c.writeUnit(base+"-"+unit.Digest, unit.Proto); err != nil {
+				return fmt.Errorf("writing unit: %v", err)
+			} else if fd, err := c.kzipFiles(r, unit); err != nil {
+				return fmt.Errorf("extracting files: %v", err)
+			} else {
+				return c.writeFiles(fd)
+			}
+		})
+		if err != nil {
+			log.Printf("Error scanning .kzip file: %v", err)
+			hasErrors = true
+		}
+	}
+	if hasErrors {
+		return subcommands.ExitFailure
+	}
+	return subcommands.ExitSuccess
+}
+
+var marshaler = &jsonpb.Marshaler{OrigName: true}
+
+func (c *cmd) writeUnit(base string, msg proto.Message) error {
+	if c.extractDir == "" {
+		err := marshaler.Marshal(os.Stdout, msg)
+		if err == nil {
+			fmt.Println()
+		}
+		return err
+	}
+	if err := os.MkdirAll(c.extractDir, 0750); err != nil {
+		return fmt.Errorf("creating output directory: %v", err)
+	}
+	f, err := os.Create(filepath.Join(c.extractDir, base+".unit"))
+	if err != nil {
+		return err
+	}
+	err = marshaler.Marshal(f, msg)
+	cerr := f.Close()
+	if err != nil {
+		return err
+	}
+	return cerr
+}
+
+func (c *cmd) writeFiles(fd []fileData) error {
+	if c.extractDir == "" {
+		return nil
+	}
+	var g errgroup.Group
```

Comment:
`g, ctx := errgroup.WithContext(context.Background())`

Use context in `g.Go` so that things get cancelled properly on errors.